File: packages/contracts/python/noa_contracts/style_profile.py
from __future__ import annotations

from typing import Literal
from pydantic import BaseModel


class MaterialPaletteEntry(BaseModel):
    name: str
    model_config = {"extra": "allow"}


class StyleSignals(BaseModel):
    massing_tendency: str
    facade_rhythm: str
    material_palette: list[MaterialPaletteEntry]
    lighting_mood: str
    surface_language: str
    compositional_balance: Literal["symmetric", "asymmetric", "dynamic"]
    emotional_tone: list[str]
    negative_preferences: list[str]
    roof_tendency: str
    glazing_ratio: float
    overhang_tendency: Literal["deep", "minimal", "none"]


class StyleProfile(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    id: str
    inspiration_board_id: str
    inference_job_id: str
    embedding_vector: list[float]
    style_signals: StyleSignals
    confidence_per_signal: dict[str, float]
    source_image_ids: list[str]
    created_at: str

File: packages/contracts/v1/style-profile.ts
import type { UUID, ISO8601 } from "./common.js";

export type MassingTendency = string;
export type FacadeRhythm = string;
export type LightingMood = string;
export type SurfaceLanguage = string;
export type RoofTendency = string;

export interface MaterialPaletteEntry {
  name: string;
  [key: string]: unknown;
}

export interface StyleProfile {
  schema_version: "1.0";
  id: UUID;
  inspiration_board_id: UUID;
  inference_job_id: UUID;
  embedding_vector: number[];
  style_signals: {
    massing_tendency: MassingTendency;
    facade_rhythm: FacadeRhythm;
    material_palette: MaterialPaletteEntry[];
    lighting_mood: LightingMood;
    surface_language: SurfaceLanguage;
    compositional_balance: "symmetric" | "asymmetric" | "dynamic";
    emotional_tone: string[];
    negative_preferences: string[];
    roof_tendency: RoofTendency;
    glazing_ratio: number;
    overhang_tendency: "deep" | "minimal" | "none";
  };
  confidence_per_signal: Record<string, number>;
  source_image_ids: UUID[];
  created_at: ISO8601;
}
